solved_exercises/dani/04-functions.py

```diff
@@ -79,4 +76,143 @@ def is_prime(number):
     return all_mod_multiplied > 0
 
 
-print("Primes list up to", value, "is", primes(value))
+def test_factorial():
+    """we'll test the correctnes of factorial function"""
+
+    print("Testing factorial() function...")
+
+    test_data = [
+        # (value, expected)
+        (-1, 1),
+        (0, 1),
+        (1, 1),
+        (2, 2),
+        (9, 362880),
+    ]
+
+    fail_count = 0
+    for value, expected in test_data:
+        actual = factorial(value)
+        result = "passed" if actual == expected else "failed"
+
+        if result == "failed":
+            print(f"\t -> Actual {actual} is different from {expected}.")
+            fail_count += 1
+
+    total_count = len(test_data)
+    passed_count = total_count - fail_count
+    print(
+        f"Ran {total_count} tests out of {passed_count} were successful"
+        f" and {fail_count} did failed."
+    )
+
+
+def test_fibonnaci():
+    """we'll test the correctnes of fibonacci function"""
+
+    print("Testing fibonacci() function...")
+
+    test_data = [
+        # (value, expected)
+        (-1, []),
+        (0, []),
+        (1, [1, 1]),
+        (2, [1, 1, 2]),
+        (3, [1, 1, 2, 3]),
+        (15, [1, 1, 2, 3, 5, 8, 13]),
+    ]
+
+    fail_count = 0
+    for value, expected in test_data:
+        # actual = fibonacci2(value)
+        actual = fibonacci(value)
+        result = "passed" if actual == expected else "failed"
+
+        if result == "failed":
+            print(f"\t -> test {value} Actual {actual} is different from {expected}.")
+            fail_count += 1
+
+    total_count = len(test_data)
+    passed_count = total_count - fail_count
+    print(
+        f"Ran {total_count} tests out of {passed_count} were successful"
+        f" and {fail_count} did failed."
+    )
+
+
+def test_flatten():
+    """we'll test the correctnes of flatten function"""
+
+    print("Testing flatten() function...")
+
+    test_data = [
+        # (value, expected)
+        (
+            ["a", 13, [-16], [56.5], [147], [18.2], [18.0, [14, 15]]],
+            ["a", 13, -16, 56.5, 147, 18.2, 18.0, 14, 15],
+        ),
+        ([], []),
+        ([1, 2], [1, 2]),
+        (1, []),
+        ("abcd", []),
+    ]
```

Comment:
I would add a some new test cases, like ``[[[2, 3, 4]]]``, and ``[2, 3, [[[]]], 4]``, and  ``[[], []]`` or ``[[[]]]``